In [6]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [7]:
accident = pd.read_csv('datasets\\accident_data.csv')

In [8]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [9]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [10]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [11]:
accident['Accident_Severity']

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [12]:
accident['Accident_Severity'].unique()

array(['Serious', 'Slight', 'Fatal'], dtype=object)

In [13]:
Serious_df =accident[accident['Accident_Severity'] =='Serious'] 

In [14]:
Serious_df

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
5       200701BS67159           Serious    18-09-2019  51.497750   
...               ...               ...           ...        ...   
660629  201091NL00686           Serious     11/2/2022  57.687933   
660645  201091NL05842           Serious    18-10-2022  60.162660   
660672  201091NM01085           Serious    29-01-2022  57.170725   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
5                          Daylight  Kensington and Chelsea  -0.192556   
...                             ...                     ...        ...   
660629                     Daylight        Shetland Islands  -4.176860   
660645                     Daylight        Shetland Islands  -1.165232   
660672                     Daylight                Highland  -3.852421   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
5                          2                   3                     Dry   
...                      ...                 ...                     ...   
660629                     1                   2                     Dry   
660645                     1                   1                     Dry   
660672                     1                   2                    Snow   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
5       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660629  Single carriageway               Rural      Fine + high winds   
660645  Single carriageway               Rural      Fine + high winds   
660672  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
5                                        Car  
...                                      ...  
660629                                   Car  
660645                                   Car  
660672                                   Car  
660677                 Motorcycle over 500cc  
660

In [15]:
Slight_df = accident[accident['Accident_Severity'] =='Slight'] 

In [16]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mean())
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean())
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])

In [17]:
accident.isnull().sum()


Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [18]:
Slight_df

Index Accident_Severity Accident Date   Latitude  \
4       200701BS66837            Slight      3/9/2019  51.488576   
8       200701BS67515            Slight    31-10-2019  51.493319   
9       200701BS67543            Slight    18-10-2019  51.484539   
13      200701BS67853            Slight     8/11/2019  51.508999   
14      200701BS67859            Slight     1/11/2019  51.485125   
...               ...               ...           ...        ...   
660671  201091NM01053            Slight    29-01-2022  57.572162   
660673  201091NM01190            Slight      2/2/2022  57.585152   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   

              Light_Conditions           District Area  Longitude  \
4        Darkness - lights lit  Kensington and Chelsea  -0.192487   
8        Darkness - lights lit  Kensington and Chelsea  -0.173572   
9                     Daylight  Kensington and Chelsea  -0.175651   
13       Darkness - lights lit  Kensington and Chelsea  -0.210701   
14       Darkness - lights lit  Kensington and Chelsea  -0.167129   
...                        ...                     ...        ...   
660671                Daylight                Highland  -3.930799   
660673                Daylight                Highland  -3.748312   
660674                Daylight                Highland  -3.467828   
660675  Darkness - no lighting                Highland  -3.809281   
660676                Daylight                Highland  -3.862727   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
4                          1                   2                     Dry   
8                          1                   2                     Dry   
9                          1                   2                     Dry   
13                         3                   2             Wet or damp   
14                         1                   1                     Dry   
...                      ...                 ...                     ...   
660671                     1                   1            Frost or ice   
660673                     2                   2             Wet or damp   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   

                 Road_Type Urban_or_Rural_Area  Weather_Conditions  \
4                      NaN               Urban                 NaN   
8       Single carriageway               Urban  Fine no high winds   
9       Single carriageway               Urban  Fine no high winds   
13      Single carriageway               Urban               Other   
14          One way street               Urban  Fine no high winds   
...                    ...                 ...                 ...   
660671  Single carriageway               Rural               Other   
660673  Single carriageway               Rural  Fine no high winds   
660674  Single carriageway               Rural  Fine no high winds   
660675  Single carriageway               Rural  Fine no high winds   
660676  Single carriageway               Rural  Fine no high winds   

                                 Vehicle_Type  
4                               Other vehicle  
8                                         Car  
9       Motorcycle over 125cc and up to 500cc  
13                                        Car  
14                                        Car  
...                                       ...  
660671    Van / Goods 3.5 tonnes mgw or under  
660673                                    Car  
660674                                    Car  
660675                                    Car  
660676                                    Car  

[563801 rows x 14 columns]

In [19]:
Fatal_df = accident[accident['Accident_Severity'] =='Fatal'] 

In [20]:
Fatal_df

Index Accident_Severity Accident Date   Latitude  \
7       200701BS67370             Fatal     3/10/2019  51.482260   
11      200701BS67747             Fatal     2/11/2019  51.499127   
20      200701BS68301             Fatal    24-10-2019  51.505996   
21      200701BS68302             Fatal    24-11-2019  51.486795   
717     200701CP00032             Fatal    30-01-2019  51.510978   
...               ...               ...           ...        ...   
660470  201091NH00529             Fatal    29-01-2022  58.475726   
660519  201091NJ01533             Fatal      2/2/2022  58.495554   
660565  201091NJ11329             Fatal    15-08-2022  57.871514   
660581  201091NJ14695             Fatal    21-10-2022  58.445475   
660627  201091NL00060             Fatal      6/1/2022  60.241195   

             Light_Conditions           District Area  Longitude  \
7       Darkness - lights lit  Kensington and Chelsea  -0.179486   
11                   Daylight  Kensington and Chelsea  -0.159652   
20                   Daylight  Kensington and Chelsea  -0.214277   
21                   Daylight  Kensington and Chelsea  -0.170375   
717                  Daylight          City of London  -0.107581   
...                       ...                     ...        ...   
660470               Daylight           Western Isles  -6.278583   
660519               Daylight                Highland  -4.451628   
660565               Daylight                Highland  -4.243081   
660581  Darkness - lights lit                Highland  -3.065535   
660627               Daylight        Shetland Islands  -1.409761   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
7                          3                   2                     Dry   
11                         1                   2                     Dry   
20                         1                   1                     Dry   
21                         5                   1                     Dry   
717                        1                   2             Wet or damp   
...                      ...                 ...                     ...   
660470                     1                   1             Wet or damp   
660519                     1                   1                    Snow   
660565                     1                   1                     Dry   
660581                     1                   1             Wet or damp   
660627                     3                   1                    Snow   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
7       Single carriageway               Urban     Fine no high winds   
11      Single carriageway               Urban     Fine no high winds   
20      Single carriageway               Urban     Fine no high winds   
21      Single carriageway               Urban     Fine no high winds   
717       Dual carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660470  Single carriageway               Rural      Fine + high winds   
660519  Single carriageway               Rural  Snowing no high winds   
660565  Single carriageway               Rural     Fine no high winds   
660581  Single carriageway               Rural     Fine no high winds   
660627  Single carriageway               Rural     Fine no high winds   

                                Vehicle_Type  
7                                        Car  
11      Bus or coach (17 or more pass seats)  
20                                       Car  
21                                       Car  
717                                      Car  
...                                      ...  
660470                                   Car  
660519                                   Car  
660565         Goods 7.5 tonnes mgw and over  
660581                                   Car  
660627   Van / Goods 3.5 tonnes mgw or under  

[8661 rows x 14 columns]

In [21]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

<h2>Adjusting Data Types</h2>

<h1>QUESTIONS</h1>

<h2>1.How many accidents are there under each severity?</h2>

In [22]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>INSIGHT:</h2>


<h3>There are 563,801 accidents under slight severity, serious with 88,217 and fatal with 8,661.</h3>

<h2>2.Which district area had the most accident report?</h2>

In [23]:
accident['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h2>INSIGHT:</h2>

<h3>Based on the data ,the district area that has the most accident report is Birmingham with a total of 1349. </h3>

<h2>3.What is the average number of casualties fall under serious accident severity?</h2>

In [24]:
accident[accident['Accident_Severity'] =='Serious']['Number_of_Casualties'].mean()

np.float64(1.46727954929322)

<h2>INSIGHT:</h2>

<h3>The result shows that serious accidents usually affect one person with the average result of 1.47.</h3>

<h2>4.Where are road accidents higher, in urban or rural areas?</h2>

In [25]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>Insight:</h2>

<h3>Based on the result road accidents are much higher in Urban areas with 421663 compared to rural areas with 238990.</h3>

<h2>5.How many slight accidents occured on frost or ice compared to snow?</h2>

In [26]:
accident.groupby(['Road_Surface_Conditions','Accident_Severity']).size()

Road_Surface_Conditions  Accident_Severity
Dry                      Fatal                  5790
                         Serious               61708
                         Slight               381049
Flood over 3cm. deep     Fatal                    23
                         Serious                 152
                         Slight                  842
Frost or ice             Fatal                   193
                         Serious                2007
                         Slight                16317
Snow                     Fatal                    35
                         Serious                 565
                         Slight                 5290
Wet or damp              Fatal                  2620
                         Serious               23785
                         Slight               160303
dtype: int64

<h2>INSIGHT:</h2>

<h3>Slight accident on frost or ice shows a large number of accidents with a total of 16317 compared to snow with 5290 accidents </h3>

<h2>6.What is the specific date recorded the most number of road accidents?</h2>

In [27]:
accident['Accident Date'].value_counts()

Accident Date
2021-02-11    685
2021-10-06    680
2019-12-06    678
2019-02-01    676
2021-12-04    667
             ... 
2022-01-09    236
2022-12-02    236
2022-04-05    231
2022-02-07    213
2022-01-10    123
Name: count, Length: 576, dtype: int64

<h2>INSIGHT:</h2>

<h3>Based on the result the date that had the most accident record was on November 30, 2019.</h3>

<h2>7. What is the most common weather does accident happened ?</h2

In [28]:
accident['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>The most common weather that accidents happened the most is with fine no high winds</h3>

<h4>8. What is the most common road surface condition during accidents?
</h4>

In [29]:
common_surface = accident["Road_Surface_Conditions"].mode()[0]
common_surface

'Dry'

<h4>Insights: According the result, it shows that the most common road surface condition during accidents are dry.
</h4>

In [30]:
light_condition_accidents = accident["Light_Conditions"].value_counts()
light_condition_accidents

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h4>Insights: Based on the result of the analysis it shows that Daylight have the highest number of recorded accidents.</h4>

<h4>9. What is the average number of casualties in urban and rural areas?</h4>

In [31]:
avg_casualties_area = accident.groupby("Urban_or_Rural_Area")["Number_of_Casualties"].mean()
avg_casualties_area

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h4>Insights: Based on the data result above we can see that the average number of casualties in the rural area are higher than the urban area.</h4>

<h4>10.what is the average number of vehicles involved in accidents across different accident severity levels?</h4>

In [32]:
vehicles_acc_severity = accident.groupby("Accident_Severity")["Number_of_Vehicles"].mean()
vehicles_acc_severity

Accident_Severity
Fatal      1.786976
Serious    1.678327
Slight     1.855864
Name: Number_of_Vehicles, dtype: float64

<h4>Insights: Based on the given results, we can clearly see that Slight level of accident severity had the highest average of vehicles involved and second to it is Serious level severity, while Fatal level of severity had the lowest average.</h4>

<h4>11. Which accident severity level has the highest average number of casualties?</h4>

In [33]:
casualties_by_severity = accident.groupby("Accident_Severity")["Number_of_Casualties"].mean()
casualties_by_severity

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h4>Insights: According to the result, it shows that Fatal severity level had the highest average number of casualties.
</h4>

<h4>12. What kind of light conditions has the most serious level of accident severity recorded?</h4>

In [34]:
severity_light = accident.groupby(["Light_Conditions", "Accident_Severity"]).size().unstack()
severity_light

Accident_Severity            Fatal  Serious  Slight
Light_Conditions                                   
Darkness - lighting unknown     68      794    5622
Darkness - lights lit         1860    19130  108345
Darkness - lights unlit         45      360    2138
Darkness - no lighting        1612     7174   28651
Daylight                      5076    60759  419045

In [35]:
severity_light.sum()

Accident_Severity
Fatal        8661
Serious     88217
Slight     563801
dtype: int64

<h4>Insights: According to the results it shows that Daylight had the most serious level of accident severity with a record of 60,759 which is about 10.87% in percentage .
</h4>

<h4>
13. Which type of vehicle has the highest number of casualties?</h4>

In [36]:
casualties_per_vehicle = accident.groupby("Vehicle_Type")["Number_of_Casualties"].size()
casualties_per_vehicle

Vehicle_Type
Agricultural vehicle                       1947
Bus or coach (17 or more pass seats)      25878
Car                                      497992
Data missing or out of range                  6
Goods 7.5 tonnes mgw and over             17307
Goods over 3.5t. and under 7.5t            6096
Minibus (8 - 16 passenger seats)           1976
Motorcycle 125cc and under                15269
Motorcycle 50cc and under                  7603
Motorcycle over 125cc and up to 500cc      7656
Motorcycle over 500cc                     25657
Other vehicle                              5637
Pedal cycle                                 197
Ridden horse                                  4
Taxi/Private hire car                     13294
Van / Goods 3.5 tonnes mgw or under       34160
Name: Number_of_Casualties, dtype: int64

<h4>Insights: According to the given results we can see that the type of vehicle with the highest number of casualties is Car with a record of 497,992.
</h4>


<h4>14. Is there a correlation between the number of vehicles and the number of casualties?</h4>

In [38]:
uk_accident_corr1 = accident["Number_of_Vehicles"].corr(accident["Number_of_Casualties"])
uk_accident_corr1

np.float64(0.22888886126927557)

<h4>
Insights: Based on the result it clearly shows that there are no correlation between the no. of vehicles and the no. of casualties.</h4>

<h4>15: Which district have the highest record of road accidents?
</h4>

In [39]:
district_accident = accident['District Area'].value_counts()
district_accident

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h4>Insights: According to the result above the district with the highest record of road accident is the district of Bermingham with a result of 13,491 road accidents.</h4>

<h4>16. Are accidents more fatal in rural or urban areas?</h4>

In [40]:
severity_area = accident.groupby(["Urban_or_Rural_Area", "Accident_Severity"]).size().unstack()
severity_area

Accident_Severity     Fatal  Serious    Slight
Urban_or_Rural_Area                           
Rural                5601.0  37312.0  196077.0
Unallocated             NaN      1.0      10.0
Urban                3060.0  50904.0  367714.0

<h4>Insights: According to the given results we can clearly see that accidents are more fatal in Rural Areas than Urban Areas.</h4>